Add LFM license and next race info

iFuSiiOnzZ opened the pull request:
Add information about the LFM license, next race and countdown to the next race.

Output example:
```
Andrei Jianu    IRON+ (B1) (7.76)    ELO 1087    Assetto Corsa Competizione (PC)    <- License and game info		|
         Coach Dave GT3 Sprint Series | #102118 | Split 1 | SOF 1087 | Drivers 1	<- Next race info		       | -> Only visible when sign up for race
                                                                     00:05:25:11	<- Count down until next race   | -> Only visible when sign up for race
```

Configuration:

- User: The LFM user identifier 
- - Default: empty
- - https://lowfuelmotorsport.com/profile/[USER_IDENTIFIER]
- Font size -> from 1 to 32 
- - Default: 10
- - Increment: 1
- Update interval -> In seconds, from 30 seconds to 3600 seconds (1 hour)
- - Default 30 seconds
- - increment: 30 seconds

iFuSiiOnzZ committed:
Add LFM license and next race info

iFuSiiOnzZ commented:
@RiddleTime As a side note: My C# knowledge is basic, from what I have understood 32-bit integers are atomic, but I think strings are not, can you confirm this? Probably `_lfmLicense` should be protected by mutex.

RiddleTime commented on Race_Element.HUD.ACC/Overlays/Pitwall/OverlayLowFuelMotorsport/LowFuelMotorsport.cs:
```diff
@@ -0,0 +1,205 @@
+        {
+            return "Session Live";
+        }
+
```
```csharp
return $"{diff:dd\\:hh\\:mm\\:ss}";
```
Would this help you here?

iFuSiiOnzZ replied:
Done! Very nice built-in feature.

RiddleTime commented on Race_Element.HUD.ACC/Overlays/Pitwall/OverlayLowFuelMotorsport/LowFuelMotorsport.cs:
```diff
@@ -0,0 +1,205 @@
+
+    private void GenerateLFMLicense(JObject lfm_object)
+    {
+        string user_vorname = lfm_object["user"]["vorname"].Value<string>();
```
It would be alright to translate the entity name in combination with camel case variable naming.
```csharp
string userFirstName = ... ;
```

iFuSiiOnzZ replied:
Done! Everything changed to camel case.

RiddleTime commented on Race_Element.HUD.ACC/Overlays/Pitwall/OverlayLowFuelMotorsport/LowFuelMotorsport.cs:
```diff
@@ -0,0 +1,205 @@
+    {
+        _lfmLicenseState = DownloadState.Downloading;
+
+        var client = new global::System.Net.Http.HttpClient();
```
```csharp
using HttpClient client = new();
```

iFuSiiOnzZ replied:
Done!

RiddleTime commented on Race_Element.HUD.ACC/Overlays/Pitwall/OverlayLowFuelMotorsport/LowFuelMotorsport.cs:
```diff
@@ -0,0 +1,205 @@
+        this.Height = (int)(bounds.Height + 1);
+        this.Width = (int)(bounds.Width + 1);
+
+        g.TextRenderingHint = global::System.Drawing.Text.TextRenderingHint.AntiAlias;
```
```csharp
using System.Drawing.Text;
...
g.TextRenderingHint = TextRenderingHint.AntiAlias;
```

iFuSiiOnzZ replied:
Done! Added "using <namespace>" at the top of the file.

RiddleTime commented on Race_Element.HUD.ACC/Overlays/Pitwall/OverlayLowFuelMotorsport/LowFuelMotorsport.cs:
```diff
@@ -0,0 +1,205 @@
+    }
+    public override void Render(Graphics g)
+    {
+        string text_to_display = _lfmLicense;
```
As mentioned later on, better to use camel case
```csharp 
string licenseText;
```

iFuSiiOnzZ replied:
Done!

RiddleTime commented on Race_Element.HUD.ACC/Overlays/Pitwall/OverlayLowFuelMotorsport/LowFuelMotorsport.cs:
```diff
@@ -0,0 +1,205 @@
+    private Font _fontFamily = null;
+
+    public LowFuelMotorsport(Rectangle rectangle) : base(rectangle, "Low Fuel Motorsport")
+    {
```
Do you want to refresh the HUD at 30 Hz or would be 1 in 4 seconds be enough?
30 Herz is the default setting.

```csharp
this.RefreshRateHz = 1 / 4f;
```

iFuSiiOnzZ replied:
Well, there are 2 refreshes. One for the server fetch and the other one for the UI. The UI it should be refreshed at least one time per seconds (the minimal unit for the countdown), if the user hasn't signed in for any race it can refresh at the same time as the server fetch (but is fancy logic, so once per second is ok hahaha).

RiddleTime commented on Race_Element.HUD.ACC/Overlays/Pitwall/OverlayLowFuelMotorsport/LowFuelMotorsport.cs:
```diff
@@ -0,0 +1,205 @@
+    Description = "Shows driver license and next race info",
+    OverlayType = OverlayType.Pitwall)]
+
+internal sealed class LowFuelMotorsport : AbstractOverlay
```
`LowFuelMotorsportOverlay`
Every override/implementation of the abstract overlay should have `Overlay` added to the end of the class name. 
This way you can rename the namespace to `RaceElement.HUD.ACC.Overlays.Pitwall.LowFuelMotorsport` for both the overlay implementation as well as the overlay configuration.

iFuSiiOnzZ replied:
Done! Added "Overlay" suffix to file and class.

RiddleTime reviewed (commented):
I've left a couple of notes in the review. https://github.com/RiddleTime/Race-Element/pull/152/files

Aside from these things I'd be glad to improve on this(perhaps in styling).
Really nice addition though!

RiddleTime commented on Race_Element.HUD.ACC/Overlays/Pitwall/OverlayLowFuelMotorsport/LowFuelMotorsport.cs:
```diff
@@ -0,0 +1,205 @@
+        }
+
+        Graphics grfx = Graphics.FromImage(new Bitmap(1, 1));
+        SizeF bounds = grfx.MeasureString(text_to_display, _fontFamily);
```
There is a Graphics context available from the Method parameter (Graphics g).
```csharp
 SizeF bounds = g.MeasureString(text_to_display, _fontFamily);
```

iFuSiiOnzZ replied:
Done!

RiddleTime commented:
> @RiddleTime As a side note: My C# knowledge is basic, from what I have understood 32-bit integers are atomic, but I think strings are not, can you confirm this? Probably `_lfmLicense` should be protected by mutex.

The String is a reference type in c# and integers for example are value types(depending on 32/64 bit) so the string should not be atomic.
You can use a lock on some kind of lock object if you like to.

Though I would prefer a different solution, which would be to make a Record class from the LFM License object as well as the next race one instead of using a string.

soonic6 commented:
maybe it is much easyer adding those drvier cards:

Big Card
https://api.lowfuelmotorsport.de/lfmlicense/81594

Small Card
https://api.lowfuelmotorsport.de/lfmlicense/81594?size=sm

Is it also possible counting track cuts or incidents which results in a car damage?
This could be helpfull?
In LFM you got 1 point for cutting the track and 4 points for Damage (per Contact with a 5seconds deadtime)

iFuSiiOnzZ committed:
Update after Code Review

- Use camel case for the variables
- Add "Overlay" sufix to class name and file name
- Use C# built in methods to generate countdown string
- Group user and fetch interval under same category (Connection)
- Use "g.MeasureString" insted of Graphics.FromImage(...).MeasureString(...)
- Maximum fetch interval is 120 seconds (based on the lfm license javascript)
- Use mutex to protect access to '_lfmLicenseText' (shared variable, but not atomic)

iFuSiiOnzZ commented:
@soonic6 It's not that easy to add that, those are HTMLs (not images). You would have to render everything from scratch (not that it's not possible, but it's a lot of work because of all the logic behind it, besides, I don't know if the result will be the same).

soonic6 commented:
> It's not that easy to add that,...

okay, was just an idea. if you want to use those images, maybe for streaming, it is very easy to add this with OBS or other streaming software.

iFuSiiOnzZ commented:
Because they have some kind of built-in browser, or this software is based on electron (web application).

@RiddleTime I accept style suggestion! :)

iFuSiiOnzZ committed:
Add job to fetch information from LFM server

The user's license information should always exist, the next race only exists when the user has registered for a race.
The overlay subscribes to the job and each time new information is obtained, the overlay is notified with the information obtained.

iFuSiiOnzZ committed:
Add a max sleep interval

If the waiting time is too long, such as seconds, minutes, etc., it will not be possible to cancel the job immediately, so instead of waiting for the delta time between the rest interval, add a maximum rest value.
If the value is negative, the waiting time will be the waiting time until the next tick, a positive value will be taken as the waiting value.

RiddleTime commented:
> Because they have some kind of built-in browser, or this software is based on electron (web application).
> 
> @RiddleTime I accept style suggestion! :)

I will add some slight refactors and add some 2d styling but for now I'll merge it as the 1.0.4.0 release will include a few other things as well.

Also I am curious which language were you using before?

RiddleTime commented:
Just a note, I reverted the AbstractLoopJob change. The abstract class is written so it tries to use the least amount of cpu so it can also be used for loop mechanisms that are high frequency.

There is an IJob interface that I use in general, which can be implemented in a maybe more sophisticated job that allows for greater control.

RiddleTime commented:
So @iFuSiiOnzZ I've done some small changes as a refactor, nothing big. 
But thanks for the refactor, if you want to discuss maybe more about the app, there is a discord: https://discord.gg/26AAEW5mUq

The changes:
https://github.com/RiddleTime/Race-Element/compare/fec491567933...d5996feca7cc
